File: src/lib/fft.ts
/* Braat, adapted from Praat
 * Copyright (C) 2026 Jocelyn Stericker <jocelyn@nettek.ca>
 * Copyright (C) 1997-2011,2025 David Weenink, Paul Boersma 2016-2018,2020
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// FFT (radix-2 Cooley-Tukey, power-of-2 only)

/** Pre-computed FFT twiddle factors and bit-reversal table. Construct once per size. */
export class FftTables {
  readonly N: number
  readonly cosTable: Float32Array
  readonly sinTable: Float32Array
  readonly bitRev: Int32Array

  constructor(N: number) {
    if (N < 2 || (N & (N - 1)) !== 0)
      throw new Error(`FFT size must be a power of 2, got ${N}`)
    this.N = N
    this.cosTable = new Float32Array(N / 2)
    this.sinTable = new Float32Array(N / 2)
    this.bitRev = new Int32Array(N)
    for (let k = 0; k < N / 2; k++) {
      const a = (-2 * Math.PI * k) / N
      this.cosTable[k] = Math.cos(a)
      this.sinTable[k] = Math.sin(a)
    }
    const bits = Math.log2(N)
    for (let i = 0; i < N; i++) {
      let rev = 0
      for (let b = 0; b < bits; b++) rev = (rev << 1) | ((i >> b) & 1)
      this.bitRev[i] = rev
    }
  }
}

/**
 * In-place complex DIT FFT (Cooley-Tukey, radix-2).
 * `re` and `im` are real and imaginary arrays, both length `tables.N`, modified in-place.
 */
export function complexFFTForward(
  re: Float32Array,
  im: Float32Array,
  tables: FftTables,
): void {
  const N = tables.N
  for (let i = 0; i < N; i++) {
    const j = tables.bitRev[i]!
    if (i < j) {
      let t = re[i]!
      re[i] = re[j]!
      re[j] = t
      t = im[i]!
      im[i] = im[j]!
      im[j] = t
    }
  }
  for (let len = 2; len <= N; len <<= 1) {
    const half = len >> 1
    const step = N / len
    for (let i = 0; i < N; i += len) {
      for (let j = 0; j < half; j++) {
        const wr = tables.cosTable[j * step]
        const wi = tables.sinTable[j * step]
        const ur = re[i + j]!,
          ui = im[i + j]!
        const vr = re[i + j + half]! * wr! - im[i + j + half]! * wi!
        const vi = re[i + j + half]! * wi! + im[i + j + half]! * wr!
        re[i + j] = ur + vr
        im[i + j] = ui + vi
        re[i + j + half] = ur - vr
        im[i + j + half] = ui - vi
      }
    }
  }
}

/**
 * In-place complex DIT IFFT (Cooley-Tukey, radix-2). Conjugate twiddle pass; caller scales by 1/N.
 * `re` and `im` are real and imaginary arrays, both length `tables.N`, modified in-place.
 */
export function complexFFTInverse(
  re: Float32Array,
  im: Float32Array,
  tables: FftTables,
): void {
  const N = tables.N
  for (let i = 0; i < N; i++) {
    const j = tables.bitRev[i]!
    if (i < j) {
      let t = re[i]!
      re[i] = re[j]!
      re[j] = t
      t = im[i]!
      im[i] = im[j]!
      im[j] = t
    }
  }
  for (let len = 2; len <= N; len <<= 1) {
    const half = len >> 1
    const step = N / len
    for (let i = 0; i < N; i += len) {
      for (let j = 0; j < half; j++) {
        const wr = tables.cosTable[j * step]!
        const wi = -tables.sinTable[j * step]!
        const ur = re[i + j]!,
          ui = im[i + j]!
        const vr = re[i + j + half]! * wr - im[i + j + half]! * wi
        const vi = re[i + j + half]! * wi + im[i + j + half]! * wr
        re[i + j] = ur + vr
        im[i + j] = ui + vi
        re[i + j + half] = ur - vr
        im[i + j + half] = ui - vi
      }
    }
  }
}

File: crates/braat-dsp/src/fft.rs
/* Braat, adapted from Praat
 * Copyright (C) 2026 Jocelyn Stericker <jocelyn@nettek.ca>
 * Copyright (C) 1997-2011,2025 David Weenink, Paul Boersma 2016-2018,2020
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 */

use std::f32::consts::PI;

pub struct FftTables {
    pub n: usize,
    pub cos_table: Vec<f32>,
    pub sin_table: Vec<f32>,
    pub bit_rev: Vec<usize>,
}

impl FftTables {
    pub fn new(n: usize) -> Self {
        assert!(n >= 2 && n.is_power_of_two(), "FFT size must be a power of 2");
        let half = n / 2;
        let mut cos_table = vec![0f32; half];
        let mut sin_table = vec![0f32; half];
        for k in 0..half {
            let a = (-2.0 * PI * k as f32) / n as f32;
            cos_table[k] = a.cos();
            sin_table[k] = a.sin();
        }
        let bits = n.trailing_zeros() as usize;
        let mut bit_rev = vec![0usize; n];
        for i in 0..n {
            let mut rev = 0usize;
            for b in 0..bits {
                rev = (rev << 1) | ((i >> b) & 1);
            }
            bit_rev[i] = rev;
        }
        Self { n, cos_table, sin_table, bit_rev }
    }
}

pub fn fft_forward(re: &mut [f32], im: &mut [f32], tables: &FftTables) {
    let n = tables.n;
    for i in 0..n {
        let j = tables.bit_rev[i];
        if i < j {
            re.swap(i, j);
            im.swap(i, j);
        }
    }
    let mut len = 2;
    while len <= n {
        let half = len >> 1;
        let step = n / len;
        let mut i = 0;
        while i < n {
            for j in 0..half {
                let wr = tables.cos_table[j * step];
                let wi = tables.sin_table[j * step];
                let ur = re[i + j];
                let ui = im[i + j];
                let vr = re[i + j + half] * wr - im[i + j + half] * wi;
                let vi = re[i + j + half] * wi + im[i + j + half] * wr;
                re[i + j] = ur + vr;
                im[i + j] = ui + vi;
                re[i + j + half] = ur - vr;
                im[i + j + half] = ui - vi;
            }
            i += len;
        }
        len <<= 1;
    }
}

pub fn fft_inverse(re: &mut [f32], im: &mut [f32], tables: &FftTables) {
    let n = tables.n;
    for i in 0..n {
        let j = tables.bit_rev[i];
        if i < j {
            re.swap(i, j);
            im.swap(i, j);
        }
    }
    let mut len = 2;
    while len <= n {
        let half = len >> 1;
        let step = n / len;
        let mut i = 0;
        while i < n {
            for j in 0..half {
                let wr = tables.cos_table[j * step];
                let wi = -tables.sin_table[j * step]; // conjugate for IFFT
                let ur = re[i + j];
                let ui = im[i + j];
                let vr = re[i + j + half] * wr - im[i + j + half] * wi;
                let vi = re[i + j + half] * wi + im[i + j + half] * wr;
                re[i + j] = ur + vr;
                im[i + j] = ui + vi;
                re[i + j + half] = ur - vr;
                im[i + j + half] = ui - vi;
            }
            i += len;
        }
        len <<= 1;
    }
}
